src/ast/mod.rs

```diff
@@ -40,232 +43,110 @@ impl Display for Id {
-            NCommand::PrintSize(name) => NCommand::PrintSize(*name),
-            NCommand::Output { file, exprs } => NCommand::Output {
-                file: file.to_string(),
-                exprs: exprs.clone(),
-            },
-            NCommand::Push(n) => NCommand::Push(*n),
-            NCommand::Pop(n) => NCommand::Pop(*n),
-            NCommand::Fail(cmd) => NCommand::Fail(Box::new(cmd.map_exprs(f))),
-            NCommand::Input { name, file } => NCommand::Input {
-                name: *name,
-                file: file.clone(),
-            },
-        }
-    }
-}
-
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub enum Schedule {
-    Saturate(Box<Schedule>),
-    Repeat(usize, Box<Schedule>),
-    Run(RunConfig),
-    Sequence(Vec<Schedule>),
 }
 
-impl Schedule {
-    pub fn saturate(self) -> Self {
-        Schedule::Saturate(Box::new(self))
-    }
-}
+pub type Schedule = GenericSchedule<Symbol, Symbol, ()>;
+pub(crate) type ResolvedSchedule = GenericSchedule<ResolvedCall, ResolvedVar, ()>;
 
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub enum NormSchedule {
-    Saturate(Box<NormSchedule>),
-    Repeat(usize, Box<NormSchedule>),
-    Run(NormRunConfig),
-    Sequence(Vec<NormSchedule>),
+pub enum GenericSchedule<Head, Leaf, Ann> {
+    Saturate(Box<GenericSchedule<Head, Leaf, Ann>>),
+    Repeat(usize, Box<GenericSchedule<Head, Leaf, Ann>>),
+    Run(GenericRunConfig<Head, Leaf, Ann>),
+    Sequence(Vec<GenericSchedule<Head, Leaf, Ann>>),
 }
 
-impl NormSchedule {
-    fn to_schedule(&self) -> Schedule {
-        match self {
-            NormSchedule::Saturate(sched) => Schedule::Saturate(Box::new(sched.to_schedule())),
-            NormSchedule::Repeat(size, sched) => {
-                Schedule::Repeat(*size, Box::new(sched.to_schedule()))
-            }
-            NormSchedule::Run(config) => Schedule::Run(config.to_run_config()),
-            NormSchedule::Sequence(scheds) => {
-                Schedule::Sequence(scheds.iter().map(|sched| sched.to_schedule()).collect())
-            }
-        }
-    }
-
-    pub fn map_run_commands(&self, f: &mut impl FnMut(&NormRunConfig) -> Schedule) -> Schedule {
-        match self {
-            NormSchedule::Run(config) => f(config),
-            NormSchedule::Saturate(sched) => {
-                Schedule::Saturate(Box::new(sched.map_run_commands(f)))
-            }
-            NormSchedule::Repeat(size, sched) => {
-                Schedule::Repeat(*size, Box::new(sched.map_run_commands(f)))
-            }
-            NormSchedule::Sequence(scheds) => Schedule::Sequence(
-                scheds
-                    .iter()
-                    .map(|sched| sched.map_run_commands(f))
-                    .collect(),
-            ),
-        }
+impl<Head, Leaf, Ann> GenericSchedule<Head, Leaf, Ann> {
+    pub fn saturate(self) -> Self {
+        GenericSchedule::Saturate(Box::new(self))
     }
 }
 
-trait ToSexp {
+pub trait ToSexp {
```

Comment:
Add some documentation here (the name is self-explanatory, but it'd be good to explain why this trait is important.